apps/simplex-chat/Simplex/Store.hs

```diff
@@ -55,33 +62,122 @@ migrations =
+  let profile = Profile {contactRef, displayName}
+   in User {userId, localContactRef = contactRef, profile, activeUser}
+
+setActiveUser :: MonadUnliftIO m => SQLiteStore -> UserId -> m ()
+setActiveUser st userId = do
+  liftIO . withTransaction st $ \db -> do
+    DB.execute_ db "UPDATE users SET active_user = 0;"
+    DB.execute db "UPDATE users SET active_user = 1 WHERE user_id = ?;" (Only userId)
+
+createDirectConnection :: MonadUnliftIO m => SQLiteStore -> UserId -> ConnId -> m ()
+createDirectConnection st userId agentConnId =
+  liftIO . withTransaction st $ \db ->
+    DB.execute
+      db
+      [sql|
+        INSERT INTO connections
+          (user_id, agent_conn_id, conn_status, conn_type) VALUES (?,?,?,?);
+      |]
+      (userId, agentConnId, ConnNew, ConnContact)
+
+createDirectContact ::
+  (MonadUnliftIO m, MonadError StoreError m) => SQLiteStore -> UserId -> Connection -> Profile -> m ()
+createDirectContact st userId Connection {connId} Profile {contactRef, displayName} =
+  liftIOEither . withTransaction st $ \db -> do
+    DB.execute db "INSERT INTO contact_profiles (contact_ref, display_name) VALUES (?, ?);" (contactRef, displayName)
+    profileId <- insertedRowId db
+    lcrSuffix <- getLcrSuffix db
+    create db profileId lcrSuffix 20
+  where
+    getLcrSuffix :: DB.Connection -> IO Int
+    getLcrSuffix db =
+      maybe 0 ((+ 1) . fromOnly) . listToMaybe
+        <$> DB.queryNamed
+          db
+          [sql|
+            SELECT lcr_suffix FROM contacts
+            WHERE user_id = :user_id AND lcr_base = :contact_ref
+            ORDER BY lcr_suffix DESC
+            LIMIT 1;
+          |]
+          [":user_id" := userId, ":contact_ref" := contactRef]
+    create :: DB.Connection -> Int64 -> Int -> Int -> IO (Either StoreError ())
+    create _ _ _ 0 = pure $ Left SEDuplicateContactRef
+    create db profileId lcrSuffix attempts = do
+      let lcr = localContactRef' lcrSuffix
+      E.try (insertUser lcr) >>= \case
+        Right () -> do
+          contactId <- insertedRowId db
+          DB.execute db "UPDATE connections SET contact_id = ? WHERE connection_id = ?" (contactId, connId)
+          pure $ Right ()
+        Left e
+          | DB.sqlError e == DB.ErrorConstraint -> create db profileId (lcrSuffix + 1) (attempts - 1)
+          | otherwise -> E.throwIO e
+      where
+        localContactRef' 0 = contactRef
+        localContactRef' n = contactRef <> T.pack ('_' : show n)
+        insertUser lcr =
+          DB.execute
+            db
+            [sql|
+              INSERT INTO contacts
+                (contact_profile_id, local_contact_ref, lcr_base, lcr_suffix, user_id) VALUES (?, ?, ?, ?, ?)
+            |]
+            (profileId, lcr, contactRef, lcrSuffix, userId)
 
 deleteContact :: MonadUnliftIO m => SQLiteStore -> UserId -> ContactRef -> m ()
 deleteContact st userId contactRef =
   liftIO . withTransaction st $ \db ->
     forM_
```

Comment:
I think this method would be better off with separate queries instead of forM_ - it saves reuse of a couple of lines but slightly complicates reading.. but I won't insist

Reply:
yes, probably, not helping much